### ISSUE TITLE
assertion failure: Replay got unrecorded signal SIGSEGV (expecting SIGILL)

### ISSUE THREAD

**Issue by bjorn3** [rr-debugger/rr#2809]
The full crash session: https://gist.github.com/bjorn3/3b7dc412becb7736486b52a370d4a67f

The pack: [rr_crash.zip](https://github.com/rr-debugger/rr/files/5958713/rr_crash.zip)

This issue is deterministic. The executable uses Intel Process Tracing but due to the bug I was trying to debug using rr it crashes way before PT is started. In fact it crashes before main inside a .init_array initializer.

**Comment by khuey** [comment]
Hmm, it replays fine here.

Can you run `rr rerun --singlestep=event,icount,rip /path/to/trace` and (with compression is necessary) post that somewhere?

**Comment by bjorn3** [comment]
[rr_crash_singlestep.zip](https://github.com/rr-debugger/rr/files/5959711/rr_crash_singlestep.zip)

**Comment by khuey** [comment]
When you ran the `rr rerun` did that crash with the assertion? Your log is complete and matches mine here.

**Comment by bjorn3** [comment]
The log I uploaded is the full contents of stdout. Stderr didn't show anything, so it must not have crashed.

**Comment by khuey** [comment]
But `rr replay` crashes reliably? Does `rr replay -a` crash too?

**Comment by bjorn3** [comment]
`rr replay -a` works. `rr replay` crashes reliably when continuing to the place where the traced program crashes and then stepping back a single instruction. (`cont` + `rsi`)

**Comment by khuey** [comment]
Ahhh I missed the `rsi` in there. Ok I can reproduce this.

**Comment by khuey** [comment]
The unexpected SIGSEGV is a SEGV_ACCERR which is connected to weird hardware stuff so this is almost certainly related to using Intel PT.

**Comment by khuey** [comment]
My guess would be the PT stuff doesn't fork properly so when we go to resume from a forked checkpoint during reverse execution everything dies. Not sure we can do much here.

**Comment by bjorn3** [comment]
At the point of the crash PT shouldn't be used yet.

**Comment by khuey** [comment]
Hmm, ok. So looking again.

```[FATAL /home/khuey/dev/rr/src/ReplaySession.cc:1137:emulate_deterministic_signal()]                                                                                                           (task 63544 (rec:5985) at time 368)
 -> Assertion `t->stop_sig() == sig' failed to hold. Replay got unrecorded signal SIGSEGV (expecting SIGILL)
{signo:SIGSEGV,errno:SUCCESS,code:SEGV_ACCERR,addr:0x7f7cf6ed1d12}```

That address is in the mapping for libykshim.so.

**Comment by bjorn3** [comment]
Maybe it would be helpful to know that the tracee crashed as I messed up the `__tls_get_addr` call to jump 4 bytes earlier, thus ending up at `__fxstat64@plt+12`. I believe this is in the middle of an instruction.

**Comment by khuey** [comment]
Yeah the actual crash is definitely at a bad %rip where we've ended up in the middle of an instruction.

**Comment by Keno** [comment]
This get really confusing when jumping in the middle of an instruction, because any set breakpoints in the vicinity can change what the instruction means. We could try to detect this kind of situation and retry with all breakpoints replaced by hardware watchpoints, but that'd be kinda fancy.
